src/game/AI/ScriptDevAI/scripts/outland/hellfire_citadel/shattered_halls/shattered_hallsScripts.cpp

```diff
@@ -0,0 +1,298 @@
+    SAY_AGGRO1 = 16697,
+    SAY_AGGRO2 = 16698,
+    SAY_AGGRO3 = 16699,
+    SAY_AGGRO4 = 16700,
+    SAY_AGGRO5 = 16701,
+    SAY_AGGRO6 = 16702,
+    SAY_AGGRO7 = 16703,
+
+    SPELL_DOUBLE_ATTACK = 19818,
+
+    NPC_SHATTERED_HAND_GLADIATIOR = 17464,
+    NPC_SHATTERED_HAND_CENTURION = 17465,
+
+    MAX_GLADIATORS = 3,
+};
+
+enum GladiatiorActions
+{
+    GLADIATOR_STOP_EVENT,
+    GLADIATIOR_EVENT_RP,
+    GLADIATOR_ACTION_MAX,
+};
+
+struct npc_shattered_hand_gladiator : public CombatAI
+{
+    npc_shattered_hand_gladiator(Creature* creature) : CombatAI(creature, GLADIATOR_ACTION_MAX)
+    {
+        AddTimerlessCombatAction(GLADIATOR_STOP_EVENT, true);
+        AddCustomAction(GLADIATIOR_EVENT_RP, true, [&]() { DoDelayedReset(); });
+    }
+
+    ObjectGuid m_sparringPartner;
+
+    void Reset() override
+    {
+        CombatAI::Reset();
+        m_eventStarted = false;
+
+        DoCastSpellIfCan(nullptr, SPELL_DOUBLE_ATTACK, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT);
+    }
+
+    bool m_eventStarted;
+
+    void DoDelayedReset()
+    {
+        EnterEvadeMode();
+        DisableTimer(GLADIATIOR_EVENT_RP);
+    }
+
+    void HandleEventStart(Creature* partner)
+    {
+        m_creature->SetFactionTemporary(1693, TEMPFACTION_RESTORE_RESPAWN | TEMPFACTION_RESTORE_COMBAT_STOP);
+        AttackStart(partner);
+        m_sparringPartner = partner->GetObjectGuid();
+
+        SendAIEvent(AI_EVENT_CUSTOM_C, m_creature, partner);
+    }
+
+    void ReceiveAIEvent(AIEventType eventType, Unit* /*sender*/, Unit* invoker, uint32 /*miscValue*/) override
+    {
+        if (eventType == AI_EVENT_CUSTOM_A) // Attacked by Player
+        {
+            if (Creature* partner = m_creature->GetMap()->GetCreature(m_sparringPartner))
+            {
+                m_creature->getThreatManager().modifyThreatPercent(partner, -101);
+                m_creature->ClearTemporaryFaction();
+                m_eventStarted = false;
+                m_creature->SetInCombatWithZone();
+            }
+        }
+        else if (eventType == AI_EVENT_CUSTOM_B) // End Duel
+        {
+            m_creature->AttackStop();
+            m_creature->ClearTemporaryFaction();
+            m_creature->HandleEmote(EMOTE_ONESHOT_BOW);
+            ResetTimer(GLADIATIOR_EVENT_RP, 1000);
+            m_eventStarted = false;
+        }
+        else if (eventType == AI_EVENT_CUSTOM_C) // End Duel
+        {
+            m_creature->SetFactionTemporary(1692, TEMPFACTION_RESTORE_RESPAWN | TEMPFACTION_RESTORE_COMBAT_STOP);
+            AttackStart(invoker);
+            m_sparringPartner = invoker->GetObjectGuid();
+        }
+    }
+    void ExecuteAction(uint32 action) override
+    {
+        switch (action)
+        {
+        case GLADIATOR_STOP_EVENT:
+        {
+            if (m_creature->GetHealthPercent() <= 40.f)
+            {
+                if (Creature* centurion = GetClosestCreatureWithEntry(m_creature, NPC_SHATTERED_HAND_CENTURION, 15.f, true, false, true))
```

Comment:
If i were you, I would strongly suggest storing the ObjectGuid of the npc he is sparring with and not doing a search. Players can get up to some weird shenanigans where they pull the wrong one to proximity and you get a weird edgecase bug somehow :D The script in front of hellfire citadel has this exact issue with bugging out.

Reply:
Not sure how best to do this but there should be no way this can happen.

The Centurion is only the npc who starts/stops the fight. and if players start the fight with the centurion, or the gladiators they should all attack the players